core/arch/arm/crypto/aes-gcm-ce.c

```diff
@@ -29,23 +30,36 @@ static void put_be_block(void *dst, const void *src)
 	put_be64((uint8_t *)dst + 8, s[0]);
 }
 
+static void ghash_reflect(uint64_t h[2], const uint64_t k[2])
+{
+	uint64_t b = get_be64(k);
+	uint64_t a = get_be64(k + 1);
+
+	h[0] = (a << 1) | (b >> 63);
+	h[1] = (b << 1) | (a >> 63);
+	if (b >> 63)
+		h[1] ^= 0xc200000000000000UL;
+}
+
 void internal_aes_gcm_set_key(struct internal_aes_gcm_state *state,
 			      const struct internal_aes_gcm_key *enc_key)
 {
 	uint64_t k[2];
-	uint64_t a;
-	uint64_t b;
+	uint64_t h[2];
 
 	crypto_aes_enc_block(enc_key->data, sizeof(enc_key->data),
 			     enc_key->rounds, state->ctr, k);
 
-	/* Store hash key in little endian and multiply by 'x' */
-	b = get_be64(k);
-	a = get_be64(k + 1);
-	state->ghash_key.k[0] = (a << 1) | (b >> 63);
-	state->ghash_key.k[1] = (b << 1) | (a >> 63);
-	if (b >> 63)
-		state->ghash_key.k[1] ^= 0xc200000000000000UL;
+	ghash_reflect(state->ghash_key.h, k);
+
+	internal_aes_gcm_gfmul(k, k, h);
+	ghash_reflect(state->ghash_key.h2, h);
+
+	internal_aes_gcm_gfmul(k, h, h);
+	ghash_reflect(state->ghash_key.h3, h);
+
+	internal_aes_gcm_gfmul(k, h, h);
+	ghash_reflect(state->ghash_key.h4, h);
```

Comment:
for commit "core: update AArch64 GHASH acceleration routines":
This function does more that original one without caller of `internal_aes_gcm_set_key()`. I guess this is due to change in the assemble loop. Maybe this should be highlighted with some details in the commit message. Maybe simply a reference like https://github.com/projectacrn/acrn-kernel/commit/22240df7ac6d76a271197571a7be45addef2ba15.

Reply:
Good idea, I'll add such a link in the commit message.